app/dao/jobs_dao.py

```diff
@@ -148,16 +155,35 @@ def dao_get_jobs_older_than_data_retention(notification_types):
     return jobs
 
 
-def dao_get_all_letter_jobs():
-    return db.session.query(
-        Job
-    ).join(
-        Job.template
-    ).filter(
-        Template.template_type == LETTER_TYPE,
-        # test letter jobs (or from research mode services) are created with a different filename,
-        # exclude them so we don't see them on the send to CSV
-        Job.original_file_name != LETTER_TEST_API_FILENAME
-    ).order_by(
-        desc(Job.created_at)
+@transactional
+def dao_cancel_letter_job(job):
+    number_of_notifications_cancelled = Notification.query.filter(
+        Notification.job_id == job.id
+    ).update({'status': NOTIFICATION_CANCELLED,
+              'updated_at': datetime.utcnow(),
+              'billable_units': 0})
+    job.job_status = JOB_STATUS_CANCELLED
+    dao_update_job(job)
+    return number_of_notifications_cancelled
+
+
+def can_letter_job_be_cancelled(job):
+    template = dao_get_template_by_id(job.template_id)
+    if template.template_type != LETTER_TYPE:
+        return "Only letter jobs can be cancelled through this endpoint. This is not a letter job."
+
+    notifications = Notification.query.filter(
+        Notification.job_id == job.id
     ).all()
+    count_notifications = len(notifications)
+    if job.job_status != JOB_STATUS_FINISHED or count_notifications != job.notification_count:
+        return "This job is still being processed. Wait a couple of minutes and try again."
+    count_cancellable_notifications = len([
+        n for n in notifications if n.status in CANCELLABLE_JOB_LETTER_STATUSES
+    ])
+    if count_cancellable_notifications != job.notification_count or not letter_can_be_cancelled(
+        NOTIFICATION_CREATED, job.created_at
```

Comment:
We should also consider if the job has been scheduled

Reply:
good point, for scheduled jobs we should go through `cancel_job` endpoint, right? But then we shouldn't show the "cancel letter job" link in admin to avoid confusion, I will make sure this doesn't happen 👍